src/cls/rbd/cls_rbd_types.cc

```diff
@@ -316,5 +317,138 @@ bool GroupSpec::is_valid() const {
+  ::decode(group_id, it);
+  ::decode(snapshot_id, it);
+  DECODE_FINISH(it);
+}
+
+void GroupSnapshotNamespace::dump(Formatter *f) const {
+  f->dump_int("group_pool", group_pool);
+  f->dump_string("group_id", group_id);
+  f->dump_string("snapshot_id", snapshot_id);
+}
+
+class EncodeSnapshotNamespaceVisitor : public boost::static_visitor<void> {
+public:
+  explicit EncodeSnapshotNamespaceVisitor(bufferlist &bl) : m_bl(bl) {
+  }
+
+  template <typename T>
+  inline void operator()(const T& t) const {
+    ::encode(static_cast<uint32_t>(T::SNAPSHOT_NAMESPACE_TYPE), m_bl);
+    t.encode(m_bl);
+  }
+
+private:
+  bufferlist &m_bl;
+};
+
+class DecodeSnapshotNamespaceVisitor : public boost::static_visitor<void> {
+public:
+  DecodeSnapshotNamespaceVisitor(bufferlist::iterator &iter)
+    : m_iter(iter) {
+  }
+
+  template <typename T>
+  inline void operator()(T& t) const {
+    t.decode(m_iter);
+  }
+private:
+  bufferlist::iterator &m_iter;
+};
+
+class DumpSnapshotNamespaceVisitor : public boost::static_visitor<void> {
+public:
+  explicit DumpSnapshotNamespaceVisitor(Formatter *formatter, const std::string &key)
+    : m_formatter(formatter), m_key(key) {}
+
+  template <typename T>
+  inline void operator()(const T& t) const {
+    auto type = T::SNAPSHOT_NAMESPACE_TYPE;
+    m_formatter->dump_string(m_key.c_str(), stringify(type));
+    t.dump(m_formatter);
+  }
+private:
+  ceph::Formatter *m_formatter;
+  std::string m_key;
+};
+
+class GetTypeVisitor : public boost::static_visitor<SnapshotNamespaceType> {
+public:
+  template <typename T>
+  inline SnapshotNamespaceType operator()(const T&) const {
+    return static_cast<SnapshotNamespaceType>(T::SNAPSHOT_NAMESPACE_TYPE);
+  }
+};
+
+
+SnapshotNamespaceType SnapshotNamespaceEntry::get_namespace_type() const {
+  return static_cast<SnapshotNamespaceType>(boost::apply_visitor(GetTypeVisitor(),
+								 snapshot_namespace));
+}
+
+void SnapshotNamespaceEntry::encode(bufferlist& bl) const {
+  ENCODE_START(1, 1, bl);
+  boost::apply_visitor(EncodeSnapshotNamespaceVisitor(bl), snapshot_namespace);
+  ENCODE_FINISH(bl);
+}
+
+void SnapshotNamespaceEntry::decode(bufferlist::iterator &p)
+{
+  DECODE_START(1, p);
+  uint32_t snap_type;
+  ::decode(snap_type, p);
+  switch (snap_type) {
+    case cls::rbd::SNAPSHOT_NAMESPACE_TYPE_USER:
+      snapshot_namespace = UserSnapshotNamespace();
+      break;
+    case cls::rbd::SNAPSHOT_NAMESPACE_TYPE_GROUP:
+      snapshot_namespace = GroupSnapshotNamespace();
+      break;
+    default:
+      snapshot_namespace = UnknownSnapshotNamespace();
+      break;
+  }
+  boost::apply_visitor(DecodeSnapshotNamespaceVisitor(p), snapshot_namespace);
+  DECODE_FINISH(p);
+}
+
+void SnapshotNamespaceEntry::dump(Formatter *f) const {
+  boost::apply_visitor(DumpSnapshotNamespaceVisitor(f, "snapshot_namespace_type"), snapshot_namespace);
+}
+
+void SnapshotNamespaceEntry::generate_test_instances(std::list<SnapshotNamespaceEntry *> &o) {
+
```

Comment:
Random but valid is fine -- it's to help ensure that the format isn't unintentionally broken by future changes.